src/commands/git-metadata/gitdb.ts

```diff
@@ -62,14 +50,63 @@ export const uploadToGitDB = async (
 
   let commitsToExclude: string[]
   try {
-    commitsToExclude = await getKnownCommits(log, request, repoURL, latestCommits)
+    commitsToExclude = await getKnownCommits(log, request, repositoryURL, latestCommits)
     log.debug(`${commitsToExclude.length} commits already in GitDB.`)
   } catch (err) {
     log.warn(`Failed getting commits to exclude: ${err}`)
     throw err
   }
 
-  const commitsToInclude = latestCommits.filter((x) => !commitsToExclude.includes(x))
+  return {
+    commitsToInclude: latestCommits.filter((x) => !commitsToExclude.includes(x)),
+    commitsToExclude,
+    headCommit: latestCommits[0],
+  }
+}
+
+export const uploadToGitDB = async (
+  log: Logger,
+  request: RequestBuilder,
+  git: simpleGit.SimpleGit,
+  dryRun: boolean,
+  repositoryURL?: string
+) => {
+  let repoURL
+  if (repositoryURL) {
+    repoURL = repositoryURL
+  } else {
+    try {
+      repoURL = await getRepoURL(git)
+      log.debug(`Syncing repository ${repoURL}`)
+    } catch (err) {
+      log.warn(`Failed getting repository URL: ${err}`)
+      throw err
+    }
+  }
+
+  let commitsToInclude: string[]
+  let commitsToExclude: string[]
+  let headCommit: string
+
+  const getCommitsBeforeUnshallowing = await getCommitsToInclude(log, request, git, repoURL)
+
+  commitsToInclude = getCommitsBeforeUnshallowing.commitsToInclude
+  commitsToExclude = getCommitsBeforeUnshallowing.commitsToExclude
+  headCommit = getCommitsBeforeUnshallowing.headCommit
+
+  // If there are no commits to include, it means the backend already has all the commits.
+  if (commitsToInclude.length === 0) {
+    return
+  }
+  // If there are commits to include and the repository is shallow, we need to repeat the process after unshallowing
+  const isShallow = await isShallowRepository(git)
+  if (isShallow) {
+    await unshallowRepository(log, git)
+    const getCommitsAfterUnshallowing = await getCommitsToInclude(log, request, git, repoURL)
```

Comment:
Should we check the repo `isShallow` before invoking to the backend the first time? 

If I understood correctly, with the current approach if in the first iteration the found local commits is an incomplete subset, the backend might returned that all the commits are known, but.. in case we were unshallow the repo before, the local commits subset might be bigger (?), and therefore, the backend might return that some commits are missing. 
Also, it looks we're doing the work twice. 

Assuming that `isShallow` method is performant enough, should we do it always in order to reduce the back and force with the backend and have a more accurate set of commits store in the backend?

I've seen we're changing the tests explicitly to implement this use case. Just throwing a challenge here, because if looks unexpected to me.

Am I missing something?

Reply:
> Assuming that isShallow method is performant enough

It's not, that's why we're doing this (indeed weird) dance.

The assumption here is that even if the clone is shallow, you'll at least have _one_ commit, and that'll be the last one. If the backend does have this last commit, we assume that it has the rest, so there's no need for unshallowing. If it doesn't we indeed need to unshallow and check again (so getting latest commit and requesting `search_commit` again).